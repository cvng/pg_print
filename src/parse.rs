use crate::fmt::Printer;
use crate::stmt::RawStmt;

impl Printer {
    #[cfg(feature = "unstable")]
    pub fn parse(&mut self, parse: &parser::Parse) {
        self.cbox(0);
        for stmt in &parse.stmts {
            self.stmt(&RawStmt { stmt: &stmt.stmt });
        }
        self.end();
    }

    #[cfg(not(feature = "unstable"))]
    pub fn parse(&mut self, parse: &pg_query::protobuf::ParseResult) {
        self.cbox(0);
        for stmt in &parse.stmts {
            if let Some(stmt) = &stmt.stmt {
                if let Some(node) = &stmt.node {
                    self.stmt(&RawStmt { stmt: node });
                }
            }
        }
        self.end();
    }
}
