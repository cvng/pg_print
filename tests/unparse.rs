use insta::assert_snapshot;
use parser::parse_source;
use pg_query::parse;
use std::fs;
use std::path::Path;

#[test]
fn unparse() {
    let path = Path::new("tests/unparse_tests.sql");

    for (mut line, case) in fs::read_to_string(path).unwrap().lines().enumerate() {
        line += 1;

        if case.starts_with("--") {
            continue;
        }

        let unparsed = pg_print::unparse(&parse_source(case));
        let deparsed = pg_query::deparse(&parse(case).unwrap().protobuf).unwrap();
        let reparsed = pg_query::deparse(&parse(&unparsed).unwrap().protobuf).unwrap();

        let fingerprint = pg_query::fingerprint(case).unwrap();
        let description = format!("{}:{}", path.display(), line);

        assert_eq!(deparsed, reparsed, "{}", &description);
        assert_snapshot!(fingerprint.hex, unparsed, &description);
    }
}
