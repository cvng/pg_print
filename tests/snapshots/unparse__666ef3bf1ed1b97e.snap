---
source: tests/unparse.rs
expression: "tests/unparse_tests.sql:133"
---
create schema myschema
